File: src/main/java/com/yangyh/flink/java/sample/wordcount/StreamWordCount.java
package com.yangyh.flink.java.sample.wordcount;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.tuple.Tuple;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

/**
 * @description: 单词统计-stream by Java
 * @author: yangyh
 * @create: 2020-11-03 23:26
 */
public class StreamWordCount {

    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment streamEnv = StreamExecutionEnvironment.getExecutionEnvironment();

        // nk -lk 8888
        DataStreamSource<String> streamSource = streamEnv.socketTextStream("node4", 8888);


        SingleOutputStreamOperator<String> wordStream = streamSource.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public void flatMap(String line, Collector<String> collector) {
                String[] words = line.split(" ");
                for (String word : words) {
                    collector.collect(word);
                }

            }
        });

        SingleOutputStreamOperator<Tuple2<String, Integer>> wordAndCount = wordStream.map(new MapFunction<String, Tuple2<String, Integer>>() {
            @Override
            public Tuple2<String, Integer> map(String word) {

                return new Tuple2<>(word, 1);
            }
        });

        KeyedStream<Tuple2<String, Integer>, Tuple> tuple2TupleKeyedStream = wordAndCount.keyBy(0);
        SingleOutputStreamOperator<Tuple2<String, Integer>> sum = tuple2TupleKeyedStream.sum(1);
        sum.print();

        streamEnv.execute("单词统计");


    }
}

File: src/main/scala/com/yangyh/flink/scala/wc/StreamWordCount.scala
package com.yangyh.flink.scala.wc

import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment

/**
 * 流处理-单词计数
 */
object StreamWordCount {

  def main(args: Array[String]): Unit = {
    // 1.创建流计算环境
    val env = StreamExecutionEnvironment.getExecutionEnvironment

    // 2.导入隐式转换
    import org.apache.flink.streaming.api.scala._

    // 3.读取数据 Linux上启动socket流命令：nc -lk 8888
    val socketData: DataStream[String] = env.socketTextStream("node4", 8888)

    // 4.转换和处理数据
    val wordAndCount: DataStream[(String, Int)] = socketData.flatMap(_.split(" "))
      .map((_, 1))
      .keyBy(0) // 分组算子
      .sum(1) // 聚合累加算子

    // 5.打印结果(Sink)
    wordAndCount.print()

    // 6.启动流计算程序
    env.execute("流处理单词计数")
  }
}
